### ISSUE TITLE
Managed coordinates force "compile" scope

### ISSUE THREAD

**Issue by c2nes** [clj-commons/pomegranate#125]
In `cemerick.pomegranate.aether/dependency` the dependency scope is defaulted to "compile". I think this is correct for regular dependencies, but for managed dependencies this forces the scope for the dependency to be managed. For consistency with Maven, I think scope should default to "" for managed coordinates.

Here's a few examples with Leiningen and `lein with-profile -base deps :tree`,

```clojure
(defproject test "0-SNAPSHOT"
  :profiles {:dev {:dependencies [[clj-time "0.14.0"]]}})

;; [clj-time "0.14.0" :scope "test"]
;;   [joda-time "2.9.7" :scope "test"]
```

```clojure
(defproject test "0-SNAPSHOT"
  :managed-dependencies [[joda-time "2.9.7"]]
  :profiles {:dev {:dependencies [[clj-time "0.14.0"]]}})

;; [clj-time "0.14.0" :scope "test"]
;;   [joda-time "2.9.7"]
```

```clojure
(defproject test "0-SNAPSHOT"
  :managed-dependencies [[joda-time "2.9.7" :scope ""]]
  :profiles {:dev {:dependencies [[clj-time "0.14.0"]]}})

;; [clj-time "0.14.0" :scope "test"]
;;   [joda-time "2.9.7" :scope "test"]
```

**Comment by lread** [comment]
This issue is almost 3 years old with no activity.
I'm going to close it as stale for now.
If you think this needs addressing please do re-open.

**Comment by marcobiscaro2112** [comment]
I'm also affected by this (but I can't re-open the issue).

Repro without leiningen:
```
(->> {:coordinates '[[org.apache.httpcomponents.client5/httpclient5 "5.0" :scope "test"]]}
     cemerick.pomegranate.aether/resolve-dependencies*
     .getArtifactResults
     (map #(str (.getDependencyNode (.getRequest %)))))

; correct: no managed coordinates, all transitive dependencies are also scope test
=>
("org.apache.httpcomponents.client5:httpclient5:jar:5.0 (test)"
 "org.apache.httpcomponents.core5:httpcore5:jar:5.0 (test)"
 "org.apache.httpcomponents.core5:httpcore5-h2:jar:5.0 (test)"
 "org.slf4j:slf4j-api:jar:1.7.25 (test)"
 "commons-codec:commons-codec:jar:1.13 (test)")
```

```
(->> {:coordinates '[[org.apache.httpcomponents.client5/httpclient5 "5.0" :scope "test"]]
      :managed-coordinates '[[org.apache.httpcomponents.core5/httpcore5 "5.2.4"]]}
     cemerick.pomegranate.aether/resolve-dependencies*
     .getArtifactResults
     (map #(str (.getDependencyNode (.getRequest %)))))

; wrong: no dependency should be scope compile, it should just resolve the httpcore5 version without affecting the scope
=>
("org.apache.httpcomponents.client5:httpclient5:jar:5.0 (test)"
 "org.apache.httpcomponents.core5:httpcore5:jar:5.2.4 (compile)"
 "org.apache.httpcomponents.core5:httpcore5-h2:jar:5.0 (test)"
 "org.slf4j:slf4j-api:jar:1.7.25 (test)"
 "commons-codec:commons-codec:jar:1.13 (test)")
```

The implication here is that libraries can't properly use managed-dependencies in leiningen, because dev-only dependencies (that should have scope = test) end up having scope = compile, and leak in the dependency tree of users.

I can confirm that removing the default [here](https://github.com/clj-commons/pomegranate/blob/593f420772ac42abcf7525da80accb123117a4ea/src/main/clojure/cemerick/pomegranate/aether.clj#L298) solves the issue.

**Comment by marcobiscaro2112** [comment]
Opened https://github.com/clj-commons/pomegranate/pull/183 for discussion.

**Comment by lread** [comment]
Hi @marcobiscaro2112! Sorry for the delay. I've been out of commission with an injury.

I see you are a contributor to leiningen, so assume that this change is necessary to fix an issue in leiningen.

**Comment by lread** [comment]
I've started to take a peek.

So, from what I understand, `scope` differs from other managed dependency attributes in that it is the only attribute not also applied to transitive deps. A dependency inherits a managed dependency `scope` but a dependency's `scope` is always applied to it's transitive deps.

I'm surprised this bug has sat so long without folks screaming for it to be fixed. 
Is that weird?

Current tests are lacking around managed coordinates and scope; I'll beef them up.
I'll refer to #183 but likely create a new PR (I'll credit you as co-author @marcobiscaro2112 and probably ask you for a review).

**Comment by marcobiscaro2112** [comment]
Hey @lread, thanks for looking at this.

Sharing some scenarios, comparing Maven and Leiningen (via Pomegranate) behavior:

1. No scopes

1.1. Maven

```xml
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.clojure</groupId>
                <artifactId>clojure</artifactId>
                <version>1.11.1</version>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>org.clojure</groupId>
            <artifactId>data.json</artifactId>
            <version>2.5.0</version>
        </dependency>
    </dependencies>
```

```
$ mvn dependency:tree

[INFO] \- org.clojure:data.json:jar:2.5.0:compile
[INFO]    \- org.clojure:clojure:jar:1.11.1:compile
[INFO]       +- org.clojure:spec.alpha:jar:0.3.218:compile
[INFO]       \- org.clojure:core.specs.alpha:jar:0.2.62:compile
```

1.2. Leiningen

```clojure
  :managed-dependencies [[org.clojure/clojure "1.11.1"]]
  :dependencies [[org.clojure/data.json "2.5.0"]]
```

```
$ lein deps :tree

 [org.clojure/data.json "2.5.0"]
   [org.clojure/clojure "1.11.1"]
     [org.clojure/core.specs.alpha "0.2.62"]
     [org.clojure/spec.alpha "0.3.218"]
```

2. Scope only in direct dependency

2.1. Maven

```xml
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.clojure</groupId>
                <artifactId>clojure</artifactId>
                <version>1.11.1</version>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>org.clojure</groupId>
            <artifactId>data.json</artifactId>
            <version>2.5.0</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
```

```
$ mvn dependency:tree

[INFO] \- org.clojure:data.json:jar:2.5.0:test
[INFO]    \- org.clojure:clojure:jar:1.11.1:test
[INFO]       +- org.clojure:spec.alpha:jar:0.3.218:test
[INFO]       \- org.clojure:core.specs.alpha:jar:0.2.62:test
```

2.2. Leiningen

```clojure
  :managed-dependencies [[org.clojure/clojure "1.11.1"]]
  :dependencies [[org.clojure/data.json "2.5.0" :scope "test"]]
```

```
$ lein deps :tree

 [org.clojure/data.json "2.5.0" :scope "test"]
   [org.clojure/clojure "1.11.1"]
     [org.clojure/core.specs.alpha "0.2.62"]
     [org.clojure/spec.alpha "0.3.218"]
```

3. Scope only in managed transitive dependency

3.1. Maven

```xml
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.clojure</groupId>
                <artifactId>clojure</artifactId>
                <version>1.11.1</version>
                <scope>provided</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>org.clojure</groupId>
            <artifactId>data.json</artifactId>
            <version>2.5.0</version>
        </dependency>
    </dependencies>
```

```
$ mvn dependency:tree

[INFO] \- org.clojure:data.json:jar:2.5.0:compile
[INFO]    \- org.clojure:clojure:jar:1.11.1:provided
[INFO]       +- org.clojure:spec.alpha:jar:0.3.218:provided
[INFO]       \- org.clojure:core.specs.alpha:jar:0.2.62:provided
```

3.2. Leiningen

```clojure
  :managed-dependencies [[org.clojure/clojure "1.11.1" :scope "provided"]]
  :dependencies [[org.clojure/data.json "2.5.0"]]
```

```
$ lein deps :tree

 [org.clojure/data.json "2.5.0"]
   [org.clojure/clojure "1.11.1" :scope "provided"]
     [org.clojure/core.specs.alpha "0.2.62" :scope "provided"]
     [org.clojure/spec.alpha "0.3.218" :scope "provided"]
```

4. Scope in both places

4.1. Maven

```xml
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.clojure</groupId>
                <artifactId>clojure</artifactId>
                <version>1.11.1</version>
                <scope>provided</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>org.clojure</groupId>
            <artifactId>data.json</artifactId>
            <version>2.5.0</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
```

```
$ mvn dependency:tree

[INFO] \- org.clojure:data.json:jar:2.5.0:test
[INFO]    \- org.clojure:clojure:jar:1.11.1:provided
[INFO]       +- org.clojure:spec.alpha:jar:0.3.218:provided
[INFO]       \- org.clojure:core.specs.alpha:jar:0.2.62:provided
```

4.2. Leiningen

```clojure
  :managed-dependencies [[org.clojure/clojure "1.11.1" :scope "provided"]]
  :dependencies [[org.clojure/data.json "2.5.0" :scope "test"]]
```

```
$ lein deps :tree

 [org.clojure/data.json "2.5.0" :scope "test"]
   [org.clojure/clojure "1.11.1" :scope "provided"]
     [org.clojure/core.specs.alpha "0.2.62" :scope "provided"]
     [org.clojure/spec.alpha "0.3.218" :scope "provided"]
```

**Comment by marcobiscaro2112** [comment]
So, to clarify, a transitive dependency that is managed can have a scope specified in the managed dependencies (scenarios 3 and 4), and when this happens the managed dependency scope should be respected. This is already the case.

The issue happens specifically on scenario 2, where the managed dependencies _does not_ have a scope, but the direct dependencies _does_ have a scope. In this case, the direct dependency scope should be inherited by the transitive dependency, but Pomegranate force the scope of the transitive dependency to always be `"compile"`.

Does it make sense?

**Comment by lread** [comment]
Excellent and thorough examples, @marcobiscaro2112; thanks!

Yeah, scenario 2 makes sense to me, and I agree that the current behavior needs to be corrected.

Scenarios 3 and 4 are helpful, too. Initially, I thought the dependency scope should override the transitive managed dependency's specified scope, but I have learned from experimentation that this is not the case. I'll play around a wee bit more to see if I can discover any nuances.

I like your PR, but I am exploring an alternative that does not assume a `compile` scope when the scope is not specified. From the maven resolver API's perspective, an unspecified scope is valid. For a dependency, an unspecified scope is effectively `compile`, and for a transitive managed dependency, as you've expressed, an unspecified scope inherits from the dependency.

**Comment by lread** [comment]
@marcobiscaro2112, I've played around a bit more. 

I like your PR approach of explicitly defaulting to `compile` scope for direct dependencies better than the alternative I explored. Your approach is clearer and more intentional and, therefore, less prone to error and confusion. It also matches, I think, the spirit of the Maven resolver API.

I'll follow up soon with a PR that is basically the same as yours but with a few more tests.
